pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -205,10 +205,18 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 	}
 
 	for _, pod := range dswp.podManager.GetPods() {
-		if dswp.podStateProvider.ShouldPodContainersBeTerminating(pod.UID) {
+		// Keep consistency of adding pod during reconstruction
+		if dswp.hasAddedPods && dswp.podStateProvider.ShouldPodContainersBeTerminating(pod.UID) {
 			// Do not (re)add volumes for pods that can't also be starting containers
 			continue
 		}
+
+		if !dswp.hasAddedPods && dswp.podStateProvider.ShouldPodRuntimeBeRemoved(pod.UID) {
+			// When kubelet restarts, we need to add pods to dsw if there is a possibility
+			// that the container may still be running
+			continue
+		}
+
```

Comment:
tbh, i am still confused with all different ShouldPodxxx.
How about CouldHaveRunningContainers? Is it safe to use it here to guarantee no container should be running at this point when delete from dsw and trigger unmount?

Reply:
I read more comments, it is recommended to use ShouldPodxxx.
How about just use ShouldPodRuntimeBeRemoved without if part? It makes code more easy to understand.

Reply:
We ran into some quota issues before for related operations, that is why we have a more restricted condition with the dswp.hasAddedPods so we can reduce the dswp.processPodVolumes calls.

Reply:
Could you provide more details about this quota issues? This is the first time I heard about this. Thank you!

Reply:
It's not a quota issue, it's a performance issue of volume manager processing volumes it doesn't need to. Here's one example: https://github.com/kubernetes/kubernetes/issues/105204#issuecomment-1022921843

There was also some race condition that may also be impacted: https://github.com/kubernetes/kubernetes/issues/101911#issuecomment-842837401

Reply:
Sorry i used the wrong word, as Michelle said that is the issue.